File: mobile/test/budgets/budget_card_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:decimal/decimal.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:spendable/budgets/budget_card.dart';
import 'package:spendable/money.dart';
import 'package:spendable_api/spendable_api.dart';

/// The same table `lib/spendable_web/utils/budget_card_test.exs` runs, so a change to either
/// side that the other did not follow fails here.
final _fixtures = jsonDecode(File('../shared/budget_cards.json').readAsStringSync()) as Map<String, dynamic>;

const _bars = {'under': CardBar.under, 'over': CardBar.over, 'goal': CardBar.goal};

void main() {
  for (final entry in _fixtures['cards'] as List) {
    final fixture = entry as Map<String, dynamic>;
    final expected = fixture['card'] as Map<String, dynamic>;
    final source = fixture['budget'] as Map<String, dynamic>;

    test(fixture['name'] as String, () {
      final budget = Budget(
        (builder) => builder
          ..id = 'bgt_fixture'
          ..name = 'Fixture'
          ..type = BudgetTypeEnum.valueOf(source['type'] as String)
          ..balance = source['balance'] as String
          ..budgetedAmount = source['budgeted_amount'] as String?,
      );

      final card = BudgetCard.build(
        budget: budget,
        spent: money(fixture['spent'] as String),
        currentMonth: fixture['current_month'] as bool,
      );

      expect(card.label, expected['label']);
      expect(card.percent, expected['percent']);
      expect(card.bar, _bars[expected['bar']]);
      expect(card.footer, expected['footer']);
      expect(card.amount, money(expected['amount'] as String));
    });
  }

  for (final entry in _fixtures['currency'] as List) {
    final fixture = entry as Map<String, dynamic>;
    final amount = fixture['amount'] as String?;

    test('formats ${amount ?? 'nothing'} as ${fixture['formatted']}', () {
      expect(formatCurrency(amount == null ? null : Decimal.parse(amount)), fixture['formatted']);
    });
  }
}

File: lib/spendable_web/utils/budget_card_test.exs
defmodule SpendableWeb.Utils.BudgetCardTest do
  use ExUnit.Case, async: true

  import Spendable.Utils
  import SpendableWeb.Utils.BudgetCard

  alias Spendable.Budgets.Schemas.Budget

  # The iOS client reaches these same answers in Dart, so the table is what keeps the two honest.
  # mobile/test/budgets/budget_card_test.dart reads the same file.
  @fixtures "shared/budget_cards.json" |> File.read!() |> Jason.decode!()

  for fixture <- @fixtures["cards"] do
    @fixture fixture

    test @fixture["name"] do
      budgeted_amount = @fixture["budget"]["budgeted_amount"]
      funding_amount = @fixture["budget"]["funding_amount"]

      budget = %Budget{
        type: String.to_existing_atom(@fixture["budget"]["type"]),
        balance: Decimal.new(@fixture["budget"]["balance"]),
        budgeted_amount: budgeted_amount && Decimal.new(budgeted_amount),
        funding_amount: funding_amount && Decimal.new(funding_amount)
      }

      month =
        Map.new(@fixture["month"], fn {figure, amount} ->
          {String.to_existing_atom(figure), Decimal.new(amount)}
        end)

      card = build_budget_card(budget, month, @fixture["current_month"])

      assert card.label == @fixture["card"]["label"]
      assert card.percent == @fixture["card"]["percent"]
      assert card.bar == @fixture["card"]["bar"]
      assert card.footer == @fixture["card"]["footer"]
      assert Decimal.equal?(card.amount, Decimal.new(@fixture["card"]["amount"]))
    end
  end

  for fixture <- @fixtures["currency"] do
    @currency fixture

    test "formats #{inspect(fixture["amount"])} as #{fixture["formatted"]}" do
      amount = @currency["amount"] && Decimal.new(@currency["amount"])

      assert format_currency(amount) == @currency["formatted"]
    end
  end
end
